demosplan/DemosPlanCoreBundle/Logic/Statement/NonAuthorizedCaseworkerRemover.php

```diff
@@ -0,0 +1,156 @@
+        $this->procedureService = $procedureService;
+        $this->entityFetcher = $entityFetcher;
+        $this->conditionFactory = $conditionFactory;
+        $this->messageBag = $messageBag;
+        $this->entityContentChangeService = $entityContentChangeService;
+    }
+
+    public function removeNonAuthorizedCaseWorkers(string $procedureId): void
+    {
+        try {
+            $procedure = $this->procedureService->getProcedureWithCertainty($procedureId);
+            $statementsToUnsetCaseworker = $this->getStatementsToUnsetCaseworker($procedure);
+
+            if ([] !== $statementsToUnsetCaseworker) {
+                $removedCaseWorkers = $this->unsetCaseWorkers($statementsToUnsetCaseworker);
+                $this->messageBag->add(
+                    'confirm',
+                    'procedure_update.caseworker_autoremove.success',
+                    ['removedUsers' => count($removedCaseWorkers)]
+                );
+            }
+        } catch (Exception $e) {
+            $this->messageBag->add(
+                'error',
+                'procedure_update.caseworker_autoremove.error'
+            );
+        }
+    }
+
+    /**
+     * Gets an array of unauthorized caseworkers who are associated with statements in this procedure.
+     *
+     * @return array<int, Statement>
+     *
+     * @throws PathException
+     * @throws ORMException
+     * @throws OptimisticLockException
+     * @throws TransactionRequiredException
+     */
+    private function getStatementsToUnsetCaseworker(Procedure $procedure): array
+    {
+        $authorizedUserNames = $this->getAuthorizedUserNames($procedure);
+
+        return $this->entityFetcher->listEntitiesUnrestricted(
+            Statement::class,
+            [
+                $this->conditionFactory->propertyHasNotValue('', ['meta', 'caseWorkerName']),
+                $this->conditionFactory->propertyHasNotAnyOfValues(
+                    $authorizedUserNames,
+                    ['meta', 'caseWorkerName']
+                ),
+                $this->conditionFactory->propertyHasValue($procedure->getId(), ['procedure', 'id']),
+            ]
+        );
+    }
+
+    /**
+     * Gets an array of all users authorized for this procedure.
+     *
+     * @return array<int, string>
+     *
+     * @throws ORMException
+     * @throws OptimisticLockException
+     * @throws TransactionRequiredException
+     */
+    private function getAuthorizedUserNames(Procedure $procedure): array
+    {
+        $ownsProcedureCondition = $this->procedureAccessEvaluator->getOwnsProcedureCondition($procedure);
+        $authorizedUsers = $this->procedureService->getAuthorizedUsers($procedure->getId());
+        $owningUsers = $this->entityFetcher->listEntitiesUnrestricted(User::class, [$ownsProcedureCondition]);
+
+        return $authorizedUsers
+            ->merge($owningUsers)
+            ->map(static fn (User $user): string => $user->getName())
```

Comment:
I can guess why you need to use the name, but this will result in problems. Having two case workers with the same name is sadly likely enough that it will happen at some point.

Reply:
Sadly, the caseworker is literally just a string on the statementMeta. So we can only go by the name. The alternative would be to remove caseworkers in the current implementation and re-implement it as a relationship to user. But I don't even know if there isn't a use case where names of persons get added as caseworkers who aren't even users on the plattform.

Reply:
As long as the case worker feature is needed the best way for the user would be to keep the string field but to also allow the usage of a relationship _instead_ (e.g. via dropdown). When a user is removed from the procedure we can automatically remove the case worker relationships.

For the string case worker field I don't think we have another choice but to simply warn the planner on procedure settings edit that there are `n` case workers that he may want to remove manually.

Reply:
In the process of creating the first attempt to fix this (#436), I asked the PO via the related Ticket, how it should be.
I interpreted the answer like: The name should be removed in case of the user with this name will be removed from the list of authorized users.
I've take this as implicate "ok" to the resulting string comparison and therefore for the possible false positives.

But in order to take over responsibility for the code I wrote, I think you are right.
This kind of logic is to fragile. The wohle "caseworker"-thing should be questioned.
Also I doubt that the automatically removing the name of a caseworker once was a working feature of the system, which means it is a feature request, but not a bugfix.